doc/gsmpb.man
.\" -*- eval: (nroff-mode) -*-
.de TQ
.br
.ns
.TP \\$1
..
.\" Like TP, but if specified indent is more than half
.\" the current line-length - indent, use the default indent.
.de Tp
.ie \\n(.$=0:((0\\$1)*2u>(\\n(.lu-\\n(.iu)) .TP
.el .TP "\\$1"
..
.TH GSMPB 8 "##DATE##" "gsmpb v##VERSION##"
.SH NAME
gsmpb \- GSM mobile phone phonebook manipulation program
.SH SYNOPSIS
.B gsmpb
[ \fB-b\fP \fIbaudrate\fP ]
[ \fB--baudrate\fP \fIbaudrate\fP ]
[ \fB-c\fP ]
[ \fB--copy\fP ]
[ \fB-d\fP \fIdestination device or file\fP ]
[ \fB--destination\fP \fIdestination device or file\fP ]
[ \fB-h\fP ]
[ \fB--help\fP ]
[ \fB-i\fP ]
[ \fB--index\fP ]
[ \fB-I\fP \fIinit string\fP ]
[ \fB--init\fP \fIinit string\fP ]
[ \fB-p\fP \fIphonebook name\fP ]
[ \fB--phonebook\fP \fIphonebook name\fP ]
[ \fB-s\fP \fIsource device or file\fP ]
[ \fB--source\fP \fIsource device or file\fP ]
[ \fB-v\fP ]
[ \fB--version\fP ]
[ \fB-V\fP ]
[ \fB--verbose\fP ]
[ \fB-X\fP ]
[ \fB--xonxoff\fP ]
[ \fB-y\fP ]
[ \fB--synchronize\fP ]
.PP
.SH DESCRIPTION
\fIgsmpb\fP can store or retrieve phonebook entries residing in a GSM
mobile phone's phonebook to or from a file. A synchronization mode is
also available.
.PP
\fIgsmpb\fP reads entries from the source which can be a mobile phone
(if a serial device file is given) or a file (if a file name is
given). The source is never modified. \fIgsmpb\fP writes phonebook
entries to a destination file or device.  Depending on the mode the
source is copied to the destination file, thus overwriting the
destination, or the destination is synchronized with regard to the
source which is the default (details see below).
.PP
If "-" is given as the parameter for the \fB--source\fP or
\fB--destination\fP options, the phonebook is read from standard input 
and/or written to standard output, respectively.
.PP
Phonebook entries names are encoded using the GSM default alphabet in
the mobile phone, whereas they are stored using the Latin-1 encoding
in phonebook files. When reading phonebook entries from a mobile phone 
entry names are converted from the GSM default to Latin-1. Characters
that can not be converted to Latin-1 are encoded as character code
172 (Latin-1 boolean "not"). When writing file-based phonebook entries 
to a mobile phone a conversion to the GSM default alphabet takes
place. Characters that can not be converted are encoded as GSM delta
(code 16).
.PP
Error messages are printed to the standard error output. If the program
terminates on error the error code 1 is returned.
.PP
.SH OPTIONS
.TP .7i
\fB-b\fP \fIbaudrate\fP, \fB--baudrate\fP \fIbaudrate\fP
The baud rate to use. The default baudrate is 38400.
.TP .7i
\fB-c\fP, \fB--copy\fP
This causes the contents of the source to be copied to the
destination. After this operation the destination has exactly the same 
contents as the source.
.TP .7i
\fB-d\fP \fIdestination\fP, \fB--destination\fP \fIdestination\fP
The destination device or file.
.TP .7i
\fB-h\fP, \fB--help\fP
Prints an option summary.
.TP .7i
\fB-I\fP \fIinit string\fP, \fB--init\fP \fIinit string\fP
Initialization string to send to the TA (default: "E0"). Note that the
sequence "ATZ" is sent first.
.TP .7i
\fB-i\fP, \fB--index\fP
If the index position is given, \fIgsmpb\fP preserves the assignment
of entries to memory slots in the mobile phone's phonebook. This can
be used to backup phonebook entries with their position into a
phonebook file or to change the position of entries by editing a
phonebook file and writing them back to the mobile phone.
If this option is given the phonebook file used as the source
must contain indices for every entry. Additionally, these indices must 
be unique, ie. it is not allowed to assign one entry twice to a
specific position in the mobile phone's phonebook.
.TP .7i
\fB-p\fP \fIphonebook\fP, \fB--phonebook\fP \fIphonebook\fP
The name of the phonebook to read from or write to. This is only used
for device sources and destinations. Commonly available phonebooks
are:
.TP .3i
.po +0.7i
.ll 5.8i
\fIFD\fP
SIM fixdialling-phonebook
.TP .3i
\fILD\fP
SIM last-dialling-phonebook
.TP .3i
\fIME\fP
ME phonebook
.TP .3i
\fIMT\fP
combined ME and SIM phonebook
.TP .3i
\fISM\fP
SIM phonebook
.TP .3i
\fITA\fP
TA phonebook
.TP .7i
.po -0.7i
.ll 6.5i
\fB-s\fP \fIsource\fP, \fB--source\fP \fIsource\fP
The source device or file.
.TP
\fB-v\fP, \fB--version\fP
Prints the program version.
.TP .7i
\fB-V\fP, \fB--verbose\fP
Prints out a detailed progress report.
.TP .7i
\fB-X\fP, \fB--xonxoff\fP
Uses software handshaking (XON/XOFF) for accessing the device.
.TP .7i
\fB-y\fP, \fB--synchronize\fP
This causes the contents of the source to be synchronized with the
destination (default). Synchronization in this context means:
.TP .2i
.po +0.7i
.ll 5.8i
\-
If the source contains an entry with a name that does not exist in the 
destination this entry is added to the destination.
.TP .2i
\-
If the source contains an entry with a name that can also be found in
the destination, the entry in the destination is overwritten (ie. the
telephone number is updated). Exception: More then one entry with the
name exists in the destination. In this case the new entry ist just added.
.TP .2i
\-
Entries in the destination that do not exist in the source are
deleted.
.PP
Note that synchronization has the following properties that differ
from copying: This algorithm does not change the location of unchanged
entries in the destination phonebook. The synchronization function
is not case-sensitive when comparing names.
.PP
.po -0.7i
.ll 6.5i
.SH PHONEBOOK FILE FORMAT 
Phonebook entries are stored in phonebook files that are meant to be
human-readable and -editable. There is one phonebook entry per line,
and each line has the format:
.PP
.nf
index|text|phone number
.fi
.PP
The fields have the following meanings:
.TP .7i
\fIindex\fP
The index of the entry which must be a positive number. The index may
also be empty. Indices can be used in conjunction with the
\fB--index\fP option to store the entry into a specific position in
the mobile phone.
.TP .7i
\fItext\fP
Descriptive text for the entry. The text may contain the special
characters '\\', '|', carriage return (ASCII code 13), or line feed
(ASCII code 10). These must be written "\\\\", "\\|", "\\r", "\\n",
respectively. The text should only contain characters that can be
encoded using the GSM default alphabet (see comments above).
.TP .7i
\fIphone number\fP
Phone numbers can only contains the digits 0-9 and the '+' sign. A '+' 
sign denotes an international number.
.PP
.SH EXAMPLES
The following invocation of \fIgsmpb\fP synchronizes the mobile phone's
SIM phonebook with the file $HOME/.phonebook:
.PP
.nf
gsmpb --synchronize -b 19200 -d /dev/mobilephone \\
      -s $HOME/.phonebook -p "SM"
.fi
.PP
.SH AUTHOR
Peter Hofmann <software@pxh.de>
.PP
.SH BUGS
Report bugs to software@pxh.de.  Include a complete, self-contained
example that will allow the bug to be reproduced, and say which
version of \fIgsmpb\fP you are using.
.PP
.SH COPYRIGHT
Copyright \(co 1999 Peter Hofmann
.LP
\fIgsmpb\fP is free software; you can redistribute it and/or modify it under
the terms of the GNU Library General Public License as published by the Free
Software Foundation; either version 2, or (at your option) any later
version.
.LP
\fIgsmpb\fP is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public License
for more details.
.LP
You should have received a copy of the GNU Library General Public License along
with \fIgsmpb\fP; see the file COPYING.  If not, write to the Free Software
Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.PP
.SH "SEE ALSO"
.BR gsminfo(7),
.BR gsmctl(1),
.BR gsmsendsms(1),
.BR gsmsmsd(8),
.BR gsmsmsstore(1).



ChangeLog
gsmlib-1.7
     - fixed bug with calculation of userData length if userDataHeader
       is present (octet count was subtracted, not septet count)

     - fixed problem with string erase() at end of PDU for Falcom A2-1

     - -t/--charset option of gsmpb did not work due to missing parameter
       of getopt_long, fixed

     - The SMS decoder/encoder can now handle alphanumeric addresses
       in the GSM default alphabet

     - set only those SMS stores that are actually needed to perform
       SMS store operation

     - Fixed signalling error 321 (Invalid memory index) when trying to
       read from empty SMS store entry

     - Fixed parsing error when reading current network operator if no
       network connection

     - Added capability to parse cell broadcast messages to gsmlib and
       the gsmsmsd program

     - Added workaround for Motorola Timeport 260 bug that doesn't correctly
       report the message status when retrieving SMS messages from the
       store

     - Added workaround for Motorola Timeport 260 that allocates index
       numbers to SMS messages linearly so that index number can be
       be larger than capacity reported by AT command

gsmlib-1.6
     - more fixes for the COPS=? return format

     - fixed putBack() behaviour in gsm_parser (don't put back if end-of-
       stream is reached)

     - added toString() function to gsm_sms_codec's Address class

     - SMS dates and times are now output in a locale-specific manner

     - fixed bug in Parser::getEol() (_eos was accidentally set to true)

     - added gsm_win32_serial module, Win32 project (VC++),  and
       Option FirstFone changes contributed by Frediano Ziglio
       <freddy77@angelfire.com>

     - fix in COM port recognition for Win32 (gsm_util)

     - renamed library files libgsm.* to libgsmme.* (now starting with
       version 1.0) because of conflict with another Debian package

     - in gsm_phonebook and gsm_store the caching of entries can now
       be disabled

     - added facilities to use other character sets for phonebooks

     - fixed workaround for Ericcson SH888 (missing service centre address)

     - fixed bug in gsm_phonebook that prevented texts with the
       character '@' to be written to to the phonebook

     - fixed nasty memory allocation bug in gsmpb/gsmsmsstore
       (automatic MeTa variable went out of scope even though used
       later)

     - fixed Y2K problem in timestamp printing

     - fixed "make dist" to include win32 files

     - added workarounds for Falcom A2-1 (autodetection and enabling by
       "export GSMLIB_FALCOM_A2_1_FIX=1", zero after PDU)

     - fixed bug that caused gsmlib to abort with an assert if a malformed
       PDU was read

gsmlib-1.5
     - adapted MeTa::getCurrentOPInfo() and MeTa::getAvailableOPInfo()
       to handle Nokia 8290 quirks

     - code to set line speed in gsm_unix_serial reinserted (it was
       accidentally removed in previous version)

     - minor changes to initialization sequence in gsm_unix_serial

     - bugfix in gsm_unix_serial.cc: readByte() == 0 does not mean no
       more bytes, but is legal value

     - additionally allowed characters "*#pwPW" and '+' at any
       position in telephone numbers

     - added environment variable GSMLIB_SH888_FIX that (if set)
       enables the gsmlib workaround for Ericsson SH888's broken SMS TPDUs

     - fixed command line parameter handling bug in gsmsmsstore

     - fixed %files section .spec file to correctly include manual pages

     - fixed some bugs in terminal line setup (gsm_unix_serial) regarding
       software/hardware handshake 

     - added new "--sca" option to SMS-related apps to set the SMS
       service centre address on the command line (useful if default is not
       set correctly in the phone)

     - removed tcflush() call in UnixSerialPort::putLine that broke
       the event system needed for gsmsmsd

gsmlib-1.4
     - more attempts to fix UNIX serial port access

     - allow custom backends for sorted phonebooks to be integrated
       into gsmlib (eg. for RDBMS or LDAP storage). Introduced a new
       module gsm_sorted_phonebook_base that contains the infrastructure for
       this.

     - Now gsmlib needs at least gcc-2.95.2 to compile correctly.

     - Implemented option to open device with software handshaking (XON/XOFF).
       The applications now have an -X option to turn this on.

     - gsmlib now contains a facility to interrupt ongoing activity in a
       controlled way. gsm_unix_serial now blocks for one second
       maximum until it checks whether it was interrupted.

     - various small bugfixes

     - added workaround for phones that omit ':' in AT command responses

     - all debugging output is now printed to stderr

     - gsm_unix_serial: new attempt to initialize modem in a more
       robust way (contributed)

     - gsm_sms: fixed handling of user data header (contributed)

gsmlib-1.3
     - fixed bug that caused gsmlib to hang if TAs don't respond
       to AT inquiries for serial number etc.

     - fixed bug with handling of CB mode AT command construction in
       gsm_me_ta.cc, setSMSRoutingToTA()

     - allow '+' as the first character of phonenumbers

     - implemented reading and writing from/to stdin/stdout in
       gsm_sorted_[sms_store|phonebook] and gsmpb/gsmsmsstore

     - fixed (hopefully) the intermittent hangup problem in
       gsm_unix_serial_port

     - Some mobile phones cannot report the status of some facility
       locks. The gsmctl program now prints out "unknown" in the
       corresponding result line from the FLSTAT
       operation if this problem occurs (instead of terminating).

     - switched off non-blocking access to serial device

gsmlib-1.2
     - gsmlib now also works with TAs that can not switch off echo

     - fixed bug that prevented gsmlib from copying SMS_DELIVER and 
       SMS_STATUS_REPORT back to the ME


apps/gsmpb.cc
// *************************************************************************
// * GSM TA/ME library
// *
// * File:    gsmpb.cc
// *
// * Purpose: phonebook management program
// *
// * Author:  Peter Hofmann (software@pxh.de)
// *
// * Created: 24.6.1999
// *************************************************************************

#ifdef HAVE_CONFIG_H
#include <gsm_config.h>
#endif
#include <gsmlib/gsm_nls.h>
#include <string>
#ifdef WIN32
#include <gsmlib/gsm_win32_serial.h>
#else
#include <gsmlib/gsm_unix_serial.h>
#include <unistd.h>
#endif
#if defined(HAVE_GETOPT_LONG) || defined(WIN32)
#include <getopt.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <gsmlib/gsm_me_ta.h>
#include <gsmlib/gsm_util.h>
#include <gsmlib/gsm_sorted_phonebook.h>
#include <iostream>

#ifdef HAVE_GETOPT_LONG
static struct option longOpts[] =
{
  {"xonxoff", no_argument, (int*)NULL, 'X'},
  {"phonebook", required_argument, (int*)NULL, 'p'},
  {"init", required_argument, (int*)NULL, 'I'},
  {"destination", required_argument, (int*)NULL, 'd'},
  {"source", required_argument, (int*)NULL, 's'},
  {"destination-backend", required_argument, (int*)NULL, 'D'},
  {"source-backend", required_argument, (int*)NULL, 'S'},
  {"baudrate", required_argument, (int*)NULL, 'b'},
  {"charset", required_argument, (int*)NULL, 't'},
  {"copy", no_argument, (int*)NULL, 'c'},
  {"synchronize", no_argument, (int*)NULL, 'y'},
  {"help", no_argument, (int*)NULL, 'h'},
  {"version", no_argument, (int*)NULL, 'v'},
  {"verbose", no_argument, (int*)NULL, 'V'},
  {"indexed", no_argument, (int*)NULL, 'i'},
  {(char*)NULL, 0, (int*)NULL, 0}
};
#else
#define getopt_long(argc, argv, options, longopts, indexptr) \
  getopt(argc, argv, options)
#endif

// insert those entries from sourcePhonebook into destPhonebook
// that are not already present in destPhonebook

void insertNotPresent(gsmlib::SortedPhonebookRef sourcePhonebook,
                      gsmlib::SortedPhonebookRef destPhonebook,
                      bool indexed, bool verbose)
{
  for (gsmlib::SortedPhonebookBase::iterator i = sourcePhonebook->begin();
       i != sourcePhonebook->end(); ++i)
  {
    std::pair<gsmlib::SortedPhonebookBase::iterator, gsmlib::SortedPhonebookBase::iterator> range;
    if (indexed)
    {
      int index = i->index();
      range = destPhonebook->equal_range(index);
    }
    else
    {
      std::string text = i->text();
      range = destPhonebook->equal_range(text);
    }

    // do nothing if the entry is already present in the destination
    bool alreadyPresent = false;
    for (gsmlib::SortedPhonebookBase::iterator j = range.first;
         j != range.second; ++j)
    {
      i->setUseIndex(indexed);
      if (i->telephone() == j->telephone())
      {
        alreadyPresent = true;
        break;
      }
    }
    // ... else insert it
    if (!alreadyPresent)
    {
      if (verbose)
      {
        std::cout << gsmlib::stringPrintf(_("inserting '%s' tel# %s"),
                                          i->text().c_str(), i->telephone().c_str());
        if (indexed)
          std::cout << gsmlib::stringPrintf(_(" (index #%d)"), i->index());
        std::cout << std::endl;
      }
      i->setUseIndex(indexed);
      destPhonebook->insert(*i); // insert
    }
  }
}

// update those entries in destPhonebook, that
// - have the same name as one entry in destPhonebook
// - but have a different telephone number
// this is only done if the name in question is unique in the destPhonebook
// the case of several entries having the same in the sourcePhonebook
// is handled - only the first is considered for updating

void updateEntries(gsmlib::SortedPhonebookRef sourcePhonebook,
                   gsmlib::SortedPhonebookRef destPhonebook,
                   bool verbose)
{
  bool firstLoop = true;
  std::string lastText;

  for (gsmlib::SortedPhonebookBase::iterator i = sourcePhonebook->begin();
       i != sourcePhonebook->end(); ++i)
  {
    std::string text = i->text();
    if (!firstLoop && text != lastText)
    {
      std::pair<gsmlib::SortedPhonebookBase::iterator,
                gsmlib::SortedPhonebookBase::iterator> range =
        destPhonebook->equal_range(text);
      
      gsmlib::SortedPhonebookBase::iterator first = range.first;
      if (first != destPhonebook->end() && range.second == ++first)
      {                         // just one text in the destPhonebook
        if (!(*range.first == *i)) // overwrite if different in destination
        {
          if (verbose)
            std::cout << gsmlib::stringPrintf(_("updating '%s' tel# %s to new tel# %s"),
                                                range.first->text().c_str(),
                                                range.first->telephone().c_str(),
                                                i->telephone().c_str())
                      << std::endl;
          
          *range.first = *i;
        }
      }
      lastText = text;
    }
    firstLoop = false;
  }
}

// the same but for indexed phonebooks

void updateEntriesIndexed(gsmlib::SortedPhonebookRef sourcePhonebook,
                          gsmlib::SortedPhonebookRef destPhonebook,
                          bool verbose)
{
  for (gsmlib::SortedPhonebookBase::iterator i = sourcePhonebook->begin();
       i != sourcePhonebook->end(); ++i)
  {
    int index = i->index();
    
    gsmlib::SortedPhonebookBase::iterator j = destPhonebook->find(index);
    
    if (j != destPhonebook->end())
    {                           // index present in the destPhonebook
      if (! (*j == *i))         // overwrite if different in destination
      {
        if (verbose)
          std::cout << gsmlib::stringPrintf(_("updating '%s' tel# %s to new tel# %s"
                                              "(index %d)"),
                                            j->text().c_str(),
                                            j->telephone().c_str(),
                                            i->telephone().c_str(), i->index())
                    << std::endl;
        *j = *i;
      }
    }
  }
}

// delete those entries from destPhonebook, that are not present
// in sourcePhonebook

void deleteNotPresent(gsmlib::SortedPhonebookRef sourcePhonebook,
                      gsmlib::SortedPhonebookRef destPhonebook,
                      bool indexed, bool verbose)
{
  gsmlib::SortedPhonebookBase::iterator i, next;
  for (next = i = destPhonebook->begin(); next != destPhonebook->end();)
  {
    i = next++; // increase iterator here because otherwise we might erase it
    std::pair<gsmlib::SortedPhonebookBase::iterator, gsmlib::SortedPhonebookBase::iterator> range;
    if (indexed)
    {
      int index = i->index();
      range = sourcePhonebook->equal_range(index);
    }
    else
    {
      std::string text = i->text();
      range = sourcePhonebook->equal_range(text);
    }
        
    bool found = false;
    for (gsmlib::SortedPhonebookBase::iterator j = range.first;
         j != range.second; ++j)
    {
      i->setUseIndex(indexed);
      if (j->telephone() == i->telephone())
      {
        found = true;
        break;
      }
    }
    if (!found)
    {
      if (verbose)
      {
        std::cout << gsmlib::stringPrintf(_("deleting '%s' tel# %s"),
                                          i->text().c_str(), i->telephone().c_str());
        if (indexed)
          std::cout << gsmlib::stringPrintf(_(" (index #%d)"), i->index());
        std::cout << std::endl;
      }
      destPhonebook->erase(i);
#ifdef BUGGY_MAP_ERASE
      deleteNotPresent(sourcePhonebook, destPhonebook, indexed, verbose);
      return;
#endif
    }
  }
}

// *** main program

int main(int argc, char *argv[])
{
  try
  {
    // handle command line options
    std::string destination;
    std::string source;
    std::string destinationBackend;
    std::string sourceBackend;
    std::string baudrate;
    bool doSynchronize = true;
    std::string phonebook;
    gsmlib::SortedPhonebookRef sourcePhonebook, destPhonebook;
    bool verbose = false;
    bool indexed = false;
    std::string initString = gsmlib::DEFAULT_INIT_STRING;
    bool swHandshake = false;
    std::string charSet;
    gsmlib::Ref<gsmlib::MeTa> sourceMeTa, destMeTa;

    int opt;
    int dummy;
    while((opt = getopt_long(argc, argv, "I:p:s:d:b:cyhvViD:S:Xt:", longOpts,
                             &dummy))
          != -1)
      switch (opt)
      {
      case 'X':
        swHandshake = true;
        break;
      case 'I':
        initString = optarg;
        break;
      case 'V':
        verbose = true;
        break;
      case 'p':
        phonebook = optarg;
        break;
      case 'd':
        destination = optarg;
        break;
      case 's':
        source = optarg;
        break;
      case 'D':
        destinationBackend = optarg;
        break;
      case 'S':
        sourceBackend = optarg;
        break;
      case 'b':
        baudrate = optarg;
        break;
      case 't':
        charSet = optarg;
        break;
      case 'c':
        doSynchronize = false;
        break;
      case 'i':
        indexed = true;
        break;
      case 'y':
        doSynchronize = true;
        break;
      case 'v':
        std::cerr << argv[0] << gsmlib::stringPrintf(_(": version %s [compiled %s]"),
                                                     VERSION, __DATE__) << std::endl;
        exit(0);
        break;
      case 'h':
        std::cerr << argv[0] << _(": [-b baudrate][-c][-d device or file][-h]"
                                  "[-I init string]\n"
                                  "  [-p phonebook name][-s device or file]"
                                  "[-t charset][-v]"
                                  "[-V][-y][-X]") << std::endl
             << std::endl
             << _("  -b, --baudrate    baudrate to use for device "
                  "(default: 38400)")
             << std::endl
             << _("  -c, --copy        copy source entries to destination")
             << std::endl
             << _("  -d, --destination sets the destination device to "
                  "connect \n"
                  "                    to, or the file to write") << std::endl
             << _("  -D, --destination-backend sets the destination backend")
             << std::endl
             << _("  -h, --help        prints this message") << std::endl
             << _("  -i, --index       takes index positions into account")
             << std::endl
             << _("  -I, --init        device AT init sequence") << std::endl
             << _("  -p, --phonebook   name of phonebook to use") << std::endl
             << _("  -s, --source      sets the source device to connect to,\n"
                  "                    or the file to read") << std::endl
             << _("  -t, --charset     sets the character set to use for\n"
                  "                    phonebook entries") << std::endl
             << _("  -S, --source-backend sets the source backend")
             << std::endl
             << _("  -v, --version     prints version and exits") << std::endl
             << _("  -V, --verbose     print detailed progress messages")
             << std::endl
             << _("  -X, --xonxoff     switch on software handshake") << std::endl
             << _("  -y, --synchronize synchronize destination with source\n"
                  "                    entries (destination is overwritten)\n"
                  "                    (see gsmpb(1) for details)")
             << std::endl << std::endl;
        exit(0);
        break;
      case '?':
        throw gsmlib::GsmException(_("unknown option"), gsmlib::ParameterError);
        break;
      }

    // check if all parameters all present
    if (destination == "" || source == "")
      throw gsmlib::GsmException(_("both source and destination must be given"),
                                 gsmlib::ParameterError);

    // start accessing source mobile phone or file
    if (sourceBackend != "")
      sourcePhonebook =
        gsmlib::CustomPhonebookRegistry::createPhonebook(sourceBackend, source);
    else if (source == "-")
      sourcePhonebook = new gsmlib::SortedPhonebook(true, indexed);
    else if (gsmlib::isFile(source))
      sourcePhonebook = new gsmlib::SortedPhonebook(source, indexed);
    else
    {
      if (phonebook == "")
        throw gsmlib::GsmException(_("phonebook name must be given"), gsmlib::ParameterError);

      sourceMeTa = new gsmlib::MeTa(new
#ifdef WIN32
              gsmlib::Win32SerialPort
#else
              gsmlib::UnixSerialPort
#endif
              (source,
               baudrate == "" ? gsmlib::DEFAULT_BAUD_RATE :
               gsmlib::baudRateStrToSpeed(baudrate), initString,
               swHandshake));
      if (charSet != "")
        sourceMeTa->setCharSet(charSet);
      sourcePhonebook =
        new gsmlib::SortedPhonebook(sourceMeTa->getPhonebook(phonebook));
    }

    // make sure destination.c_str file exists
    if (destination != "")
    {
      try
      {
        std::ofstream f(destination.c_str(), std::ios::out | std::ios::app);
      }
      catch (std::exception&)
      {
      }
    }

    // start accessing destination mobile phone or file
    if (destinationBackend != "")
      destPhonebook =
        gsmlib::CustomPhonebookRegistry::createPhonebook(destinationBackend,
                                                 destination);
    else if (destination == "-")
      destPhonebook = new gsmlib::SortedPhonebook(false, indexed);
    else if (gsmlib::isFile(destination))
      destPhonebook = new gsmlib::SortedPhonebook(destination, indexed);
    else
    {
      if (phonebook == "")
        throw gsmlib::GsmException(_("phonebook name must be given"), gsmlib::ParameterError);

      destMeTa = new gsmlib::MeTa(new 
#ifdef WIN32
              gsmlib::Win32SerialPort
#else
              gsmlib::UnixSerialPort
#endif
              (destination,
               baudrate == "" ? gsmlib::DEFAULT_BAUD_RATE :
               gsmlib::baudRateStrToSpeed(baudrate), initString,
               swHandshake));
      if (charSet != "")
        destMeTa->setCharSet(charSet);
      gsmlib::PhonebookRef destPb = destMeTa->getPhonebook(phonebook);

      // check maximum lengths of source text and phonenumber when writing to
      // mobile phone
      unsigned int maxTextLen = destPb->getMaxTextLen();
      unsigned int maxTelLen = destPb->getMaxTelephoneLen();

      for (gsmlib::SortedPhonebookBase::iterator i = sourcePhonebook->begin();
           i != sourcePhonebook->end(); ++i) {
        if (i->text().length() > maxTextLen) {
          throw gsmlib::GsmException(
              gsmlib::stringPrintf(_("text '%s' is too large to fit into destination "
                                    "(maximum size %d characters)"),
                                   i->text().c_str(), maxTextLen),
              gsmlib::ParameterError);
        } else if (i->telephone().length() > maxTelLen) {
          throw gsmlib::GsmException(
              gsmlib::stringPrintf(_("phone number '%s' is too large to fit into "
                                     "destination (maximum size %d characters)"),
                                   i->telephone().c_str(), maxTelLen),
              gsmlib::ParameterError);
        }
      }

      // read phonebook
      destPhonebook = new gsmlib::SortedPhonebook(destPb);      
    }

    // now do the actual work
    if (doSynchronize)
    {                           // synchronizing
      if (indexed)
      {
        sourcePhonebook->setSortOrder(gsmlib::ByIndex);
        destPhonebook->setSortOrder(gsmlib::ByIndex);
        // for an explanation see below
        updateEntriesIndexed(sourcePhonebook, destPhonebook, verbose);
        deleteNotPresent(sourcePhonebook, destPhonebook, true, verbose);
        insertNotPresent(sourcePhonebook, destPhonebook, true, verbose);
      }
      else
      {
        sourcePhonebook->setSortOrder(gsmlib::ByText);
        destPhonebook->setSortOrder(gsmlib::ByText);
        // the following is done to avoid superfluous writes to the TA
        // (that takes time) and keep updated (ie. telephone number changed)
        // entries at the same place
        // 1. update entries in place where just the number changed
        updateEntries(sourcePhonebook, destPhonebook, verbose);
        // 2. delete those that are not present anymore
        deleteNotPresent(sourcePhonebook, destPhonebook, false, verbose);
        // 3. insert the new ones
        insertNotPresent(sourcePhonebook, destPhonebook, false, verbose);
      }
    }
    else
    {                           // copying
      destPhonebook->clear();
      for (gsmlib::SortedPhonebookBase::iterator i = sourcePhonebook->begin();
           i != sourcePhonebook->end(); ++i)
      {
        if (verbose)
        {
          std::cout << gsmlib::stringPrintf(_("inserting '%s' tel# %s"),
                                            i->text().c_str(), i->telephone().c_str());
          if (indexed)
            std::cout << gsmlib::stringPrintf(_(" (index #%d)"), i->index());
          std::cout << std::endl;
        }
        destPhonebook->insert(*i);
      }
    }
  }
  catch (gsmlib::GsmException &ge)
  {
    std::cerr << argv[0] << _("[ERROR]: ") << ge.what() << std::endl;
    return 1;
  }
  return 0;
}
